{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Clean Cami Output\n",
    "This notebook will read gold standard taxonomic output from the CamiSim files, the convert them to the standardized relative abundance tables."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import os\n",
    "import re\n",
    "\n",
    "path = os.path.abspath(\"/Volumes/TBHD_share/valencia/pipelines/cami_marine/data/simulation_short_read\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Movement of Short Read Files\n",
    "The original short read files were in a structure with sampleID/reads/anonymous_reads.fq.gz\n",
    "\n",
    "We want to change these to sampleID.fastq.gz, since we want to know what the sample is."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "def find_and_rename_fq_files(path: str):\n",
    "    for root, dirs, files in os.walk(path):\n",
    "        for f in files:\n",
    "            if f.endswith(\".fq.gz\"):\n",
    "                joined = os.path.join(root, f)\n",
    "                new_file_name = joined.split(\"/\")[-3] + \".fastq.gz\"\n",
    "                new_file_name = (\"_\").join(new_file_name.split(\"_\")[2:])\n",
    "                \n",
    "                new_path = os.path.join(root, new_file_name)\n",
    "                print(joined)\n",
    "                print(new_path)\n",
    "\n",
    "                os.rename(joined, new_path)\n",
    "\n",
    "find_and_rename_fq_files(path)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Parsing of Gold Standard"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Return files where the name contains the pattern \"taxonomic_profile_{0-9}.txt\"\n",
    "def get_files(path, pattern):\n",
    "    files = []\n",
    "    for file in os.listdir(path):\n",
    "        if re.match(pattern, file):\n",
    "            files.append(file)\n",
    "    return files\n",
    "\n",
    "files = get_files(path, \"taxonomic_profile_[0-9].txt\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "taxonomic_profile_9.txt\n",
      "Total RA percentage: 92.69010000000002\n",
      "Number of unique species: 274\n",
      "Top 10 species by RA percentage:\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TAXPATHSN</th>\n",
       "      <th>PERCENTAGE</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>@@TAXID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>45202.0</th>\n",
       "      <td>unidentified plasmid</td>\n",
       "      <td>37.7568</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32644.0</th>\n",
       "      <td>unidentified</td>\n",
       "      <td>20.7057</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1335757.0</th>\n",
       "      <td>Spiribacter curvatus</td>\n",
       "      <td>2.3293</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>314275.0</th>\n",
       "      <td>Alteromonas mediterranea</td>\n",
       "      <td>2.1011</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29542.0</th>\n",
       "      <td>Pelobacter acetylenicus</td>\n",
       "      <td>1.9202</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54248.0</th>\n",
       "      <td>Hyperthermus butylicus</td>\n",
       "      <td>1.5043</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>133539.0</th>\n",
       "      <td>Nitrosococcus halophilus</td>\n",
       "      <td>1.3588</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>84980.0</th>\n",
       "      <td>Desulfotalea psychrophila</td>\n",
       "      <td>1.2617</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1214906.0</th>\n",
       "      <td>unidentified virus</td>\n",
       "      <td>1.2479</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62322.0</th>\n",
       "      <td>Shewanella baltica</td>\n",
       "      <td>0.8977</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                           TAXPATHSN  PERCENTAGE\n",
       "@@TAXID                                         \n",
       "45202.0         unidentified plasmid     37.7568\n",
       "32644.0                 unidentified     20.7057\n",
       "1335757.0       Spiribacter curvatus      2.3293\n",
       "314275.0    Alteromonas mediterranea      2.1011\n",
       "29542.0      Pelobacter acetylenicus      1.9202\n",
       "54248.0       Hyperthermus butylicus      1.5043\n",
       "133539.0    Nitrosococcus halophilus      1.3588\n",
       "84980.0    Desulfotalea psychrophila      1.2617\n",
       "1214906.0         unidentified virus      1.2479\n",
       "62322.0           Shewanella baltica      0.8977"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Bottom 10 species by RA percentage:\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TAXPATHSN</th>\n",
       "      <th>PERCENTAGE</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>@@TAXID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1250205.0</th>\n",
       "      <td>Gramella sp. MAR_2010_147</td>\n",
       "      <td>0.0016</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2746.0</th>\n",
       "      <td>Halomonas elongata</td>\n",
       "      <td>0.0022</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1920883.0</th>\n",
       "      <td>Donghicola sp. JLT3646</td>\n",
       "      <td>0.0026</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42444.0</th>\n",
       "      <td>Marinovum algicola</td>\n",
       "      <td>0.0034</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1850250.0</th>\n",
       "      <td>Rhodobacter sp. LPB0142</td>\n",
       "      <td>0.0050</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59600.0</th>\n",
       "      <td>Cellulophaga algicola</td>\n",
       "      <td>0.0070</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>911045.0</th>\n",
       "      <td>Pseudovibrio sp. FO-BEG1</td>\n",
       "      <td>0.0087</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>307121.0</th>\n",
       "      <td>Micromonospora krabiensis</td>\n",
       "      <td>0.0088</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2724.0</th>\n",
       "      <td>Hirschia baltica</td>\n",
       "      <td>0.0119</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1526571.0</th>\n",
       "      <td>Lacimicrobium alkaliphilum</td>\n",
       "      <td>0.0121</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                            TAXPATHSN  PERCENTAGE\n",
       "@@TAXID                                          \n",
       "1250205.0   Gramella sp. MAR_2010_147      0.0016\n",
       "2746.0             Halomonas elongata      0.0022\n",
       "1920883.0      Donghicola sp. JLT3646      0.0026\n",
       "42444.0            Marinovum algicola      0.0034\n",
       "1850250.0     Rhodobacter sp. LPB0142      0.0050\n",
       "59600.0         Cellulophaga algicola      0.0070\n",
       "911045.0     Pseudovibrio sp. FO-BEG1      0.0087\n",
       "307121.0    Micromonospora krabiensis      0.0088\n",
       "2724.0               Hirschia baltica      0.0119\n",
       "1526571.0  Lacimicrobium alkaliphilum      0.0121"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "def split_col_and_replace(df: pd.DataFrame, col_name: str, sep: str = \"|\"):\n",
    "    df[col_name] = df[col_name].str.split(sep).str[-1]\n",
    "    return df\n",
    "\n",
    "def get_summary(df: pd.DataFrame):\n",
    "    # Summarize the total RA percentage.\n",
    "    pct = df[\"PERCENTAGE\"].sum()\n",
    "    print(f\"Total RA percentage: {pct}\")\n",
    "\n",
    "    # Summarize the number of unique species.\n",
    "    unique_species = df[\"TAXPATHSN\"].unique()\n",
    "    print(f\"Number of unique species: {len(unique_species)}\")\n",
    "\n",
    "    # Get the top 10 species.\n",
    "    top_10 = df.sort_values(by=\"PERCENTAGE\", ascending=False).head(10)\n",
    "    print(\"Top 10 species by RA percentage:\")\n",
    "    display(top_10)\n",
    "\n",
    "    # Get the bottom 10 species.\n",
    "    bottom_10 = df.sort_values(by=\"PERCENTAGE\", ascending=True).head(10)\n",
    "    print(\"Bottom 10 species by RA percentage:\")\n",
    "    display(bottom_10)\n",
    "\n",
    "def make_df(files: list, rank: str):\n",
    "    for f in files:\n",
    "        print(f)\n",
    "        df = pd.read_csv(os.path.join(path, f), sep='\\t', header=3, index_col=0, usecols=[0,1,3,4])\n",
    "\n",
    "        df = df.loc[df[\"RANK\"] == \"species\"].drop(columns=[\"RANK\"])\n",
    "        split_col_and_replace(df, \"TAXPATHSN\")\n",
    "\n",
    "        # Remove rows where the percentage is 0.0000\n",
    "        df = df.loc[df[\"PERCENTAGE\"] != 0.0000]\n",
    "\n",
    "        # Get the summary.\n",
    "        get_summary(df)\n",
    "\n",
    "        break\n",
    "\n",
    "make_df(files, \"species\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.10.6 ('.pipeline_venv': venv)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.6"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "3ac10c2973d938bf4f101ae2299756abbb7e00dac649f0769819439ff384650d"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
